### PR TITLE
Adding support for AWS SDK V2 request signer

### PR DESCRIPTION
**Pull request by matelang**
### Description
This PR introduces support for a new package that allows to use the new AWS SDK V2 to sign the OpenSearch Go client's HTTP request messages according to the AWS Signature spec V4 .

I've also updated all project dependencies to latest versions for bonus maintenance.

### Issues Resolved
No issues found in tracker relating to this.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by VijayanB** [comment]
@matelang Thanks for creating PR. Will review your PR this week.

**Comment by VachaShah** [comment]
> For tracking, I've opened an Issue and PR on AWS SDK with the fix from the second commit: [aws/aws-sdk-go-v2#1728](https://github.com/aws/aws-sdk-go-v2/issues/1728) [aws/aws-sdk-go-v2#1729](https://github.com/aws/aws-sdk-go-v2/pull/1729)

Thank you @matelang! Retrying the CI

**Comment by VijayanB** [comment]
According to their documentation:
```
The v2 SDK requires a minimum version of Go 1.15.
```
I believe opensearch-go supports from go 1.11. Should we update this as well?

**Comment by matelang** [comment]
> According to their documentation:
> 
> ```
> The v2 SDK requires a minimum version of Go 1.15.
> ```
> 
> I believe opensearch-go supports from go 1.11. Should we update this as well?

I've added a new commit to address this:
https://github.com/opensearch-project/opensearch-go/pull/116/commits/6d4b420465990ccf0024ba5ff872110ce86beb9d

**Comment by matelang** [comment]
I've also fetched upstream and adapted the import path of the signer accordingly (appended v2 to import path).

File: signer/awsv2/sdkv2signer.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+			return err
+		}
+
+		// AWS SDK v2 signer will append the content length value twice
+		// and produce an incorrect signature if previously set.
+		// To avoid this, we have to remove the content-length header.
+		r.Header.Del("Content-Length")
+	}
+
+	if err := s.signer.SignHTTP(ctx, creds, r, hash, s.service, s.awsCfg.Region, t); err != nil {
```
Comment (VijayanB) [inline signer/awsv2/sdkv2signer.go:83]:
```suggestion
	return s.signer.SignHTTP(ctx, creds, r, hash, s.service, s.awsCfg.Region, t)
```

File: signer/awsv2/sdkv2signer.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		return err
+	}
+
+	return nil
+}
+
+func hexEncodedSha256OfRequest(r *http.Request) (string, error) {
```
Comment (VijayanB) [inline signer/awsv2/sdkv2signer.go:90]: How about checking if request body == nil return empty hash or return calculated hash value from this method. In that way, hashing logic is within single method. WDYT?

Comment (matelang) [reply signer/awsv2/sdkv2signer.go:90]: I tried to keep the code as close to the original AWS SDK v1 Signer implementation as possible. In AWS SDK V1 the signer is the one calculating the body hash, so this is encapsulated.

In V2 though, it is the responsibility of the caller to calculate it.

I agree that it is cleaner like that, I'll change it now.

**Comment by VachaShah** [comment]
@matelang Looks like one of the commits has a sign off missing. Can you please fix it?

**Comment by matelang** [comment]
> @matelang Looks like one of the commits has a sign off missing. Can you please fix it?

@VachaShah I think it was the merge commit, I've fixed it by squashing all the work into just 1 commit. Have not change anything else.

**Review by VachaShah — DISMISSED** [review]
Thank you @matelang for adding this!

**Review by VijayanB — DISMISSED** [review]

**Comment by VijayanB** [comment]
@matelang Thanks for adding v2 support. Unfortunately i don't know way of writing integration test for aws iam feature. Did you get chance to test v2 support manually from opensearch-go?

**Comment by matelang** [comment]
> @matelang Thanks for adding v2 support. Unfortunately i don't know way of writing integration test for aws iam feature. Did you get chance to test v2 support manually from opensearch-go? 

We already use the vendored version of the signer in production, so far it is working great for us.

**Comment by matelang** [comment]
Is there anything else I can do to get this merged in?

**Review by VijayanB — APPROVED** [review]

**Review by VachaShah — APPROVED** [review]
Thank you

**Comment by VachaShah** [comment]
Hi @matelang, can you please create an issue for the same and link this PR so it is easier to track? Also, it would be great if we can add this to the [USER_GUIDE](https://github.com/opensearch-project/opensearch-go/blob/main/USER_GUIDE.md#how-to-use-iams-as-authentication-method).

**Comment by matelang** [comment]
> Hi @matelang, can you please create an issue for the same and link this PR so it is easier to track? Also, it would be great if we can add this to the [USER_GUIDE](https://github.com/opensearch-project/opensearch-go/blob/main/USER_GUIDE.md#how-to-use-iams-as-authentication-method).

Created the issue and linked it here.

I will open another PR later today modifying the User guide accordingly. 
Thanks for your help in getting this merged in.

**Comment by matelang** [comment]
@VachaShah added example and enhanced docs:
https://github.com/opensearch-project/opensearch-go/pull/131

**Comment by dblock** [comment]
@matelang Do you know if this works for chunked gzipped requests? Care to try? Coming from https://github.com/opensearch-project/opensearch-py/issues/176 and https://github.com/acm19/aws-request-signing-apache-interceptor/issues/20

**Comment by matelang** [comment]
> @matelang Do you know if this works for chunked gzipped requests? Care to try? Coming from [opensearch-project/opensearch-py#176](https://github.com/opensearch-project/opensearch-py/issues/176) and [acm19/aws-request-signing-apache-interceptor#20](https://github.com/acm19/aws-request-signing-apache-interceptor/issues/20)

I have not tested it explicitly to be honest. We are using the default configurations for our OpenSearch client connection. I can try to reproduce it sometime time this week.

Thanks for letting me know.

**Pull request merged**